File: reproduce.py
#!/usr/bin/env -S uv run --script

import argparse
from dataclasses import dataclass, field
from pathlib import Path

import yaml

from checkout import Linux, checkout
from make import Build, build_kstep, build_linux
from run import Driver, run_qemu
from scripts import (
    PROJ_DIR,
    ResultDir,
    TermColor,
    system,
)


BUGS_FILE = PROJ_DIR / "bugs.yaml"


@dataclass(frozen=True)
class Bug:
    """One bugs.yaml entry: how to build, run, reproduce and fuzz a bug (fields documented there)."""
    name: str
    # Commit-based: buggy = fix~1, fixed = fix
    fix: str | None = None
    # Patch-based: buggy = ref, fixed = ref + patch
    ref: str | None = None
    patch: str | None = None
    # Extra kernel config fragment to merge
    config: Path | None = None
    extra: bool = False
    plot_format: str | None = None
    # The machine (run.py Driver fields)
    num_cpus: int = 2
    mem_mb: int = 128  # kSTEP itself needs ~20 MB
    # Fuzzing cli material (checks, verbs, setup), see bugs.yaml
    fuzz: dict = field(default_factory=dict)

    @property
    def machine(self) -> dict:
        return {k: getattr(self, k) for k in ("num_cpus", "mem_mb")}

    @property
    def driver(self) -> Driver:
        return Driver(name=self.name, **self.machine)

    @property
    def linux(self) -> list[Linux]:
        if self.fix:
            return [
                Linux(name="buggy", ref=f"{self.fix}~1", config=self.config),
                Linux(name="fixed", ref=self.fix, config=self.config),
            ]
        elif self.ref and self.patch:
            patch_path = PROJ_DIR / "linux" / self.patch
            return [
                Linux(name="buggy", ref=self.ref, config=self.config),
                Linux(
                    name="fixed",
                    ref=self.ref,
                    patch=patch_path,
                    config=self.config,
                ),
            ]
        else:
            raise ValueError(
                f"Bug '{self.name}': specify either 'fix' or 'ref'+'patch'"
            )

    def print_facts(self):
        """One `key value` line per fact, for the fuzzer."""
        print("name", self.name)
        for k, v in self.machine.items():
            if v is not None:
                print(k, v)
        for k in ("checks", "verbs", "setup"):
            for line in self.fuzz.get(k, []):
                print(k.rstrip("s") if k != "setup" else k, line)


def load_bugs() -> dict[str, Bug]:
    return {name: Bug(name=name, **(entry or {})) for name, entry in yaml.safe_load(BUGS_FILE.read_text()).items()}


ALL_BUGS = load_bugs()
BUGS = [b for b in ALL_BUGS.values() if not b.extra]
BUGS_EXTRA = [b for b in ALL_BUGS.values() if b.extra]


def log_step(prefix: str, message: str):
    print(f"{TermColor.GREEN}{prefix}{TermColor.RESET}: {message}", flush=True)


def plot_data(python_script: str, driver: str):
    log_step(driver, "Plotting data")
    system(f"{PROJ_DIR}/scripts/plot_{python_script}.py {driver}")


def reproduce(bug: "Bug", linux: Linux):
    kernel = f"{bug.name}_{linux.name}"

    log_step(kernel, "Checkout Linux")
    checkout(linux.ref, kernel=kernel, patch=linux.patch, tarball=True, set_current=False)
    b = Build(kernel)
    log = b.dir / "build.log"
    log_step(kernel, f"Build Linux (log: {log})")
    build_linux(b, extra_config=linux.config, log=log)
    log_step(kernel, "Build kSTEP")
    build_kstep(b)

    result_dir = ResultDir.create(f"repro_{bug.name}/{linux.name}")
    log_step(kernel, "Run kSTEP")
    log_step(kernel, f"QEMU Log: {result_dir.log}")
    log_step(kernel, f"kSTEP Output: {result_dir.output}")
    run_qemu(kernel=kernel, driver=bug.driver, result_dir=result_dir, headless=True)


def main(bug: Bug, runs: list[str]):
    print("=" * 80, flush=True)
    print(f" {bug.name} ".center(80, "="), flush=True)
    print("=" * 80, flush=True)

    linux_map = {linux.name: linux for linux in bug.linux}
    for r in runs:
        if r == "plot":
            continue
        linux = linux_map.get(r, Linux(name=r, ref=r))
        reproduce(bug, linux)

    if "plot" in runs:
        if bug.plot_format:
            plot_data(bug.plot_format, bug.name)
        else:
            print(f"Plot format not specified for bug '{bug.name}'")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "name",
        type=str,
        default="all",
        choices=["all", "extra", *ALL_BUGS],
        help="The name of the bug to reproduce, or 'all' to reproduce all BUGS.",
    )
    parser.add_argument(
        "--run",
        type=str,
        default=["buggy", "fixed", "plot"],
        nargs="+",
    )
    parser.add_argument("--print", action="store_true", help="print the bug's facts from bugs.yaml and exit (the fuzzer reads them)")
    args = parser.parse_args()

    if args.print:
        ALL_BUGS[args.name].print_facts()
        raise SystemExit

    if args.name == "all":
        selected_bugs = BUGS
    elif args.name == "extra":
        selected_bugs = BUGS_EXTRA
    else:
        selected_bugs = [ALL_BUGS[args.name]]

    print(f"running {len(selected_bugs)} bug(s)")
    for bug in selected_bugs:
        main(bug=bug, runs=args.run)

File: crates/kstep/src/cli/reproduce.rs
use anyhow::{Context, Result};
use clap::ValueEnum;
use kstep::bugs::{self, Bug, Kernel};
use kstep::cmd::{cmd, run, GREEN, RESET};
use kstep::{Build, ResultDir};

/// Check out, build and run a bug on its buggy and fixed kernels, then plot the two traces
#[derive(clap::Args)]
#[command(after_help = "Examples:
  kstep reproduce sync_wakeup            # build/sync_wakeup_{buggy,fixed} -> results/repro_sync_wakeup/
  kstep reproduce all                    # every bug in the paper's table; `extra` for the others
  kstep reproduce freeze --steps plot    # redraw from the traces already there")]
pub struct Args {
    /// A bugs.yaml entry, `all` (the paper's table) or `extra` (the rest)
    bug: String,
    /// Which parts to run
    #[arg(long, value_enum, num_args = 1.., default_values = ["buggy", "fixed", "plot"])]
    steps: Vec<Step>,
}

#[derive(Clone, Copy, PartialEq, Eq, ValueEnum)]
enum Step {
    Buggy,
    Fixed,
    Plot,
}

fn log_step(prefix: &str, message: &str) {
    println!("{GREEN}{prefix}{RESET}: {message}");
}

/// One kernel of a bug: checked out, built (kernel log in build.log), and the driver run on it.
fn reproduce(bug: &Bug, kernel: &Kernel) -> Result<()> {
    let name = bug.build_name(kernel.name);
    log_step(&name, &format!("Checkout Linux {}", kernel.git_ref));
    let b = Build::new(Some(&name))?;
    let log = b.dir().join("build.log");
    log_step(&name, &format!("Build Linux (log: {})", log.display()));
    b.build_linux(None, true, Some(&log))?;
    log_step(&name, "Build kSTEP");
    b.build_kstep()?;

    let results = ResultDir::create(Some(&format!("repro_{}/{}", bug.name, kernel.name)), false)?;
    log_step(
        &name,
        &format!(
            "Run kSTEP (log: {}, output: {})",
            results.kernel_log().display(),
            results.kstep_log().display()
        ),
    );
    let boot = b.boot(&bug.name, bug.machine(), &results, false);
    run(&mut boot.command(), None)
        .with_context(|| format!("see {}", results.kernel_log().display()))
}

pub fn main(a: Args) -> Result<()> {
    let all = bugs::load()?;
    let selected: Vec<&Bug> = match a.bug.as_str() {
        "all" => all.values().filter(|b| !b.extra).collect(),
        "extra" => all.values().filter(|b| b.extra).collect(),
        name => vec![bugs::lookup(&all, name)?],
    };
    println!("running {} bug(s)", selected.len());
    for bug in selected {
        println!("{:=^80}", format!(" {} ", bug.name));
        let [buggy, fixed] = bug.kernels()?;
        for (step, kernel) in [(Step::Buggy, &buggy), (Step::Fixed, &fixed)] {
            if a.steps.contains(&step) {
                reproduce(bug, kernel)?;
            }
        }
        if a.steps.contains(&Step::Plot) {
            match &bug.plot_format {
                Some(f) => {
                    log_step(&bug.name, "Plot");
                    run(
                        cmd(
                            "uv",
                            [
                                "run",
                                "--script",
                                &format!("scripts/plot_{f}.py"),
                                &bug.name,
                            ],
                        )
                        .current_dir(kstep::proj_dir()),
                        None,
                    )?;
                }
                None => println!("no plot format for '{}'", bug.name),
            }
        }
    }
    Ok(())
}
